Whether or not a pacifist can fire a gun is no longer based on the gun being harmful, but the chambered round being harmful

iskyp opened the pull request:
:cl:
fix: makes dragnet non harmful
tweak: pacifists can now use any disabler or stun setting on any energy gun
code: removed all of the pacifism check code from code/modules/mob/living/living.dm
code: gun objects no longer have a harmful variable, instead, ammo_casing objects now have a harmful variable, which is by default set to TRUE
code: if a pacifist fires a gun, it checks whether or not the round chambered is lethal, instead of whether or not the gun itself is lethal.
/:cl:

resolves #36600 
[video link](https://youtu.be/CkM6rn7Bz4s)

Commit:
out with the old

Commit:
in with the new

*whether or not a pacifist can fire a gun is no longer based on whether or not the gun in question is "harmful"

*it is now based on whether or not the round chambered is harmful

*works for energy weapons

*works for ballistic weapons (although i haven't actually designated any ballistic weapons as harmful

*DRAGnet is no longer harmful (unless PK get's his PR merged before this)

*Foam Dart with pens in them are now harmful because they do like 5 brute damage and that is NOT OKAY

*no, soporific rounds do not work like they did in the video, that was just me showing off

Commit:
Merge branch 'master' into pacifist

Rinoka commented:
as mentioned in the issue, it has some... debatable implications (pacifists with a gun they didn't load able to sense the lethality of the bullets)

but if we gonna merge we gonna merge, i'm just raising the potential issue so we know what we want pacifism to be

Rinoka commented:
also, wasn't sure: are pacifists able to use syringe guns

iskyp commented:
no

Rinoka commented:
ok good

those would be A Mess

Iamgoofball commented:
Knocking people down is inherently harmful

iskyp commented:
open an issue about how the trait isn't realistic enough then

Rinoka commented:
having goof _still_ not be gitbanned is inherently harmful

and yet here we are

iskyp commented:
FUCK BURST FIRE WEAPONS WORK HOLD ON

Commit:
burst fire is no longer broken

woopsie daisy i fucked it up
also changed the "hey you can't do this this gun is a bad man gun" text slightly

chaoko99 commented:
Pacifist, being one who pacifies. That actually doesn't explicitly deny killing but... It's better meant as subdued here.

AlexTheSysop42 commented:
> pacifists with a gun they didn't load able to sense the lethality of the bullets

I assume you're talking about ballistic weapons, and I don't think there are any of those that aren't harmful. (Even the detective's gun does some damage.)

iskyp commented:
soporific rounds for the .50 BMG are nonlethal but I haven't designated that because they aren't like that currently

ExcessiveUseOfCobblestone commented on code/modules/projectiles/gun.dm:
```diff
@@ -212,6 +211,10 @@
 	if(chambered && chambered.BB)
+		if(user.has_trait(TRAIT_PACIFISM))
+			if(chambered.harmful)
+				to_chat(user, "<span class='notice'> [src] is lethal! You don't want to risk harming anyone...</span>")
```
The gun doesn't matter, the bullet inside it does. The notification should explain this.

ExcessiveUseOfCobblestone commented on code/modules/projectiles/gun.dm:
```diff
@@ -257,6 +260,10 @@
 			addtimer(CALLBACK(src, .proc/process_burst, user, target, message, params, zone_override, sprd, randomized_gun_spread, randomized_bonus_spread, rand_spr, i), fire_delay * (i - 1))
 	else
 		if(chambered)
+			if(user.has_trait(TRAIT_PACIFISM))
```
shooting something should not check if you have a trait, having a certain trait should pick up when you shoot something.

Then again this is guncode so...

ExcessiveUseOfCobblestone commented on code/modules/projectiles/gun.dm:
```diff
@@ -257,6 +260,10 @@
 		if(chambered)
+			if(user.has_trait(TRAIT_PACIFISM))
+				if(chambered.harmful)
+					to_chat(user, "<span class='notice'> [src] is lethal! You don't want to risk harming anyone...</span>")
```
> The gun doesn't matter, the bullet inside it does. The notification should explain this.

ExcessiveUseOfCobblestone commented:
also you want `fixes #36600` in the PR body.

I still disagree pacifism should be able to stun at all in a stun-combat system we have currently, and believe the actual fix would be to remove the ability to do such rather than provide more outlets to subdue individuals.

AlexTheSysop42 commented:
The whole thing about a stun-combat system is that once someone's stunned, the attacker has to actually hit the them to kill them, otherwise they just get back up. Pacifists can't use the stun to kill.

iskyp commented:
@ExcessiveUseOfCobblestone 
> i was slightly upset and made #36600 then decided to fix it
also added commenting to appease you

Commit:
commenting

now you can understand the shitcode without having to have any knowledge of shitcode

ExcessiveUseOfCobblestone commented:
I wasn't saying it's unclear, I'm saying that's not the correct way to approach it if possible.

Checking if the gun is harmful when you try to shoot it should be on pacifism rather than checking if the individual has pacifism everytime you shoot the gun.

iskyp commented:
ill see what I can do when I get back home

Anonmare commented:
Going by the original pacifists, The Ai and its metal legion, tasers and disablers are not harmful so this would be fine albeit a technicality

Iamgoofball commented:
AIs aren't pacifist buddy they just don't want harm occurring to humans, they'll go full murderbot on non humans

Rinoka commented:
They still are technically the _original_ pacifists, considering all that complex nonhuman bullshit is just meta rearing its head and back in the day AIs would still consider antags human, which resulted in them requiring pacifism.

You're not wrong nowadays that we get full murderbot AIs but still

iskyp commented:
so after dicking around with code for a few days i don't know how to tie the code directly to the trait like @ExcessiveUseOfCobblestone suggested
I've gotten it to work somewhat
1) change `if(shooty && !shooty.harmful)` on line 860 of living.dm to 
`if(shooty && !shooty.chambered.harmful)`
medbeam is rendered unable to fire because it lacks an ammo_casing and breaks
2) the code i currently have
works perfectly but might be less than ideal in terms of code optimization

if someone could nudge me onto the right path if there is one that'd be pretty cool because i'm still fairly new to dm

ExcessiveUseOfCobblestone commented:
Don't worry i'll do it later

iskyp commented:
yes, pacifists can cuff

Bawhoppen commented:
No better way to prevent violence, than to lock everybody up.

Commit:
clarifies why you can't fire

just a minor "can't fire" message edit
